modal-go/task_command_router_client.go

```diff
@@ -0,0 +1,629 @@
+}
+
+// TryInitTaskCommandRouterClient attempts to initialize a TaskCommandRouterClient.
+// Returns nil if command router access is not available for this task.
+func TryInitTaskCommandRouterClient(
+	ctx context.Context,
+	serverClient pb.ModalClientClient,
+	taskID string,
+	logger *slog.Logger,
+	profile Profile,
+) (*TaskCommandRouterClient, error) {
+	resp, err := serverClient.TaskGetCommandRouterAccess(ctx, pb.TaskGetCommandRouterAccessRequest_builder{
+		TaskId: taskID,
+	}.Build())
+	if err != nil {
+		if st, ok := status.FromError(err); ok && st.Code() == codes.FailedPrecondition {
+			logger.DebugContext(ctx, "Command router access is not enabled for task", "task_id", taskID)
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	logger.DebugContext(ctx, "Using command router access for task", "task_id", taskID, "url", resp.GetUrl())
+
+	url, err := url.Parse(resp.GetUrl())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse task router URL: %w", err)
+	}
+
+	if url.Scheme != "https" {
+		return nil, fmt.Errorf("task router URL must be https, got: %s", resp.GetUrl())
+	}
+
+	host := url.Hostname()
+	port := url.Port()
+	if port == "" {
+		port = "443"
+	}
+	target := fmt.Sprintf("%s:%s", host, port)
+
+	var creds credentials.TransportCredentials
+	if profile.TaskCommandRouterInsecure {
+		logger.WarnContext(ctx, "Using insecure TLS for task command router due to MODAL_TASK_COMMAND_ROUTER_INSECURE")
+		creds = insecure.NewCredentials()
+	} else {
+		// Use custom TLS credentials that skip ALPN enforcement.
+		// The command router server may not negotiate ALPN, which causes
+		// grpc-go v1.67+ to fail the handshake.
+		creds = &tlsCredsNoALPN{}
+	}
+
+	conn, err := grpc.NewClient(
+		target,
+		grpc.WithTransportCredentials(creds),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(maxMessageSize),
+			grpc.MaxCallSendMsgSize(maxMessageSize),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create task command router connection: %w", err)
+	}
+
+	client := &TaskCommandRouterClient{
+		stub:         pb.NewTaskCommandRouterClient(conn),
+		conn:         conn,
+		serverClient: serverClient,
+		taskID:       taskID,
+		serverURL:    resp.GetUrl(),
+		logger:       logger,
+	}
+	jwt := resp.GetJwt()
+	client.jwt.Store(&jwt)
+	jwtExp := parseJwtExpiration(ctx, jwt, logger)
+	client.jwtExp.Store(jwtExp)
+
+	logger.DebugContext(ctx, "Successfully initialized command router client", "task_id", taskID)
+	return client, nil
+}
+
+// Close closes the gRPC connection.
+func (c *TaskCommandRouterClient) Close() error {
+	if !c.closed.CompareAndSwap(false, true) {
+		return nil
+	}
+	if c.conn != nil {
+		return c.conn.Close()
+	}
+	return nil
+}
+
+func (c *TaskCommandRouterClient) authContext(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer "+*c.jwt.Load())
+}
+
+func (c *TaskCommandRouterClient) refreshJwt(ctx context.Context) error {
```

Comment:
I'm a little surprised there is no long around `refreshJwt`, in principle two go-routines can refresh at the same time and atomically `c.jwt`.

An alternative is to lock this so when one go-routine updates the JWT, it'll update both   `c.jwtExp` and `c.jwt`. When the other go-routine reloads, it'll see that `c.jwtExp` is new enough and not make the `TaskGetCommandRouterAccess` RPC call.

Reply:
I believe the `refreshJwtGroup.Do` below, using [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight) should accomplish the same thing, or no?

Reply:
Ah I missed that. Yup `singleflight` should work.